File: py/wlearn/ensemble/_selection.py
"""Caruana greedy ensemble selection matching JS @wlearn/ensemble/selection.js."""

import numpy as np

from ..errors import ValidationError
from ..automl._cv import get_scorer


def caruana_select(oof_predictions, y_true, max_size=20, scoring='accuracy',
                   task='classification', n_classes=0, refine_weights=True):
    """Greedy ensemble selection (Caruana et al., 2004).

    Selects a weighted subset from a pool of OOF predictions by greedily
    adding the candidate that most improves the ensemble score at each step.
    Candidates can be selected multiple times (with replacement).

    Args:
        oof_predictions: list of np.ndarray per candidate
            Classification: each (n * n_classes,) flat row-major proba
            Regression: each (n,)
        y_true: np.ndarray
        max_size: max ensemble members
        scoring: metric name or callable
        task: 'classification' or 'regression'
        n_classes: inferred from data if 0
        refine_weights: if True, optimize weights via projected gradient descent
            after Caruana selection (requires optimize_weights from _weights.py)

    Returns:
        dict with 'indices', 'weights', 'scores'
    """
    n = len(y_true)
    n_candidates = len(oof_predictions)

    if n_candidates == 0:
        raise ValidationError('caruana_select: need at least 1 candidate')

    scorer_fn = get_scorer(scoring)

    pred_size = len(oof_predictions[0]) / n
    if pred_size != int(pred_size):
        raise ValidationError(
            'caruana_select: oof_predictions[0].length must be divisible by n'
        )
    pred_size = int(pred_size)

    if task == 'classification' and n_classes == 0:
        n_classes = pred_size

    # Current ensemble prediction (running weighted average)
    current = np.zeros(len(oof_predictions[0]), dtype=np.float64)
    selected = []
    scores = []

    for t in range(max_size):
        best_idx = -1
        best_score = -float('inf')

        for i in range(n_candidates):
            trial = _trial_predictions(current, oof_predictions[i], t, t + 1)
            trial_score = _score(trial, y_true, scorer_fn, task, n_classes, n)
            if trial_score > best_score:
                best_score = trial_score
                best_idx = i

        selected.append(best_idx)
        scores.append(best_score)

        # Update running ensemble
        P = oof_predictions[best_idx]
        for j in range(len(current)):
            current[j] = (t * current[j] + P[j]) / (t + 1)

    # Compute weights from selection counts
    counts = {}
    for idx in selected:
        counts[idx] = counts.get(idx, 0) + 1

    unique_indices = np.array(sorted(counts.keys()), dtype=np.int32)
    weights = np.array(
        [counts[int(idx)] / max_size for idx in unique_indices],
        dtype=np.float64
    )

    result = {
        'indices': unique_indices,
        'weights': weights,
        'scores': np.array(scores, dtype=np.float64),
    }

    if refine_weights and len(unique_indices) > 1:
        from ._weights import optimize_weights
        selected_oofs = [oof_predictions[int(idx)] for idx in unique_indices]
        refined = optimize_weights(
            selected_oofs, y_true, weights, task=task,
        )
        result['weights'] = refined

    return result


def _trial_predictions(current, candidate, t_count, t_total):
    trial = np.empty(len(current), dtype=np.float64)
    for j in range(len(current)):
        trial[j] = (t_count * current[j] + candidate[j]) / t_total
    return trial


def _score(preds, y_true, scorer_fn, task, n_classes, n):
    if task == 'regression':
        return scorer_fn(y_true, preds)
    # Classification: convert proba to hard predictions via argmax
    hard_preds = np.zeros(n, dtype=np.float64)
    for i in range(n):
        best_c = 0
        best_v = -float('inf')
        for c in range(n_classes):
            if preds[i * n_classes + c] > best_v:
                best_v = preds[i * n_classes + c]
                best_c = c
        hard_preds[i] = best_c
    return scorer_fn(y_true, hard_preds)

File: packages/ensemble/src/selection.js
const { getScorer, normalizeY, ValidationError } = require('@wlearn/core')
const { optimizeWeights } = require('./weights.js')

/**
 * Caruana greedy ensemble selection (Caruana et al., 2004).
 */
function caruanaSelect(oofPredictions, yTrue, {
  maxSize = 20,
  scoring = 'accuracy',
  task = 'classification',
  nClasses = 0,
  refineWeights = true,
} = {}) {
  const yn = normalizeY(yTrue)
  const n = yn.length
  const nCandidates = oofPredictions.length

  if (nCandidates === 0) {
    throw new ValidationError('caruanaSelect: need at least 1 candidate')
  }

  const scorerFn = getScorer(scoring)

  // Determine prediction size per sample
  const predSize = oofPredictions[0].length / n
  if (predSize !== Math.floor(predSize)) {
    throw new ValidationError('caruanaSelect: oofPredictions[0].length must be divisible by n')
  }

  if (task === 'classification' && nClasses === 0) {
    nClasses = predSize
  }

  // Current ensemble prediction (running weighted average)
  const current = new Float64Array(oofPredictions[0].length)
  const selected = []
  const scores = []

  for (let t = 0; t < maxSize; t++) {
    let bestIdx = -1
    let bestScore = -Infinity

    for (let i = 0; i < nCandidates; i++) {
      // Trial: ((t) * current + P[i]) / (t + 1)
      const trial = _trialPredictions(current, oofPredictions[i], t, t + 1)
      const trialScore = _score(trial, yn, scorerFn, task, nClasses, n)
      if (trialScore > bestScore) {
        bestScore = trialScore
        bestIdx = i
      }
    }

    selected.push(bestIdx)
    scores.push(bestScore)

    // Update running ensemble: current = (t * current + P[bestIdx]) / (t + 1)
    const P = oofPredictions[bestIdx]
    for (let j = 0; j < current.length; j++) {
      current[j] = (t * current[j] + P[j]) / (t + 1)
    }
  }

  // Compute weights from selection counts
  const counts = new Map()
  for (const idx of selected) {
    counts.set(idx, (counts.get(idx) || 0) + 1)
  }
  const uniqueIndices = new Int32Array([...counts.keys()].sort((a, b) => a - b))
  const weights = new Float64Array(uniqueIndices.length)
  for (let i = 0; i < uniqueIndices.length; i++) {
    weights[i] = counts.get(uniqueIndices[i]) / maxSize
  }

  const result = {
    indices: uniqueIndices,
    weights,
    scores: new Float64Array(scores),
  }

  if (refineWeights && uniqueIndices.length > 1) {
    const selectedOofs = Array.from(uniqueIndices, idx => oofPredictions[idx])
    result.weights = optimizeWeights(selectedOofs, yn, weights, { task })
  }

  return result
}

// --- Internal helpers ---

function _trialPredictions(current, candidate, tCount, tTotal) {
  const trial = new Float64Array(current.length)
  for (let j = 0; j < current.length; j++) {
    trial[j] = (tCount * current[j] + candidate[j]) / tTotal
  }
  return trial
}

function _score(preds, yTrue, scorerFn, task, nClasses, n) {
  if (task === 'regression') {
    return scorerFn(yTrue, preds)
  }
  // Classification: convert proba to hard predictions via argmax
  const hardPreds = new Float64Array(n)
  for (let i = 0; i < n; i++) {
    let bestC = 0, bestV = -Infinity
    for (let c = 0; c < nClasses; c++) {
      if (preds[i * nClasses + c] > bestV) {
        bestV = preds[i * nClasses + c]
        bestC = c
      }
    }
    hardPreds[i] = bestC
  }
  return scorerFn(yTrue, hardPreds)
}

module.exports = { caruanaSelect }
